File: controller/utils/lifecycle.py
"""
Lifecycle management for the Orchestry Controller.
Handles startup and shutdown events for all components.
"""
import logging
import threading
import time
import os
from typing import Optional, Any

from controller.manager import AppManager
from state.db import get_database_manager
from controller.nginx import DockerNginxManager
from controller.scaler import AutoScaler, ScalingPolicy
from controller.health import HealthChecker
from controller.cluster import DistributedController

logger = logging.getLogger(__name__)

# Global components - initialized when starting the API
app_manager: Optional[AppManager] = None
state_store: Optional[Any] = None
nginx_manager: Optional[DockerNginxManager] = None
auto_scaler: Optional[AutoScaler] = None
health_checker: Optional[HealthChecker] = None
cluster_controller: Optional[DistributedController] = None

# Background monitoring task
monitoring_task: Optional[threading.Thread] = None
monitoring_active = False

# Nginx request tracking to compute RPS
_prev_nginx_requests: Optional[int] = None
_prev_nginx_time: Optional[float] = None


def get_app_manager() -> Optional[AppManager]:
    """Get the global app manager instance."""
    return app_manager


def get_state_store() -> Optional[Any]:
    """Get the global state store instance."""
    return state_store


def get_nginx_manager() -> Optional[DockerNginxManager]:
    """Get the global nginx manager instance."""
    return nginx_manager


def get_auto_scaler() -> Optional[AutoScaler]:
    """Get the global auto scaler instance."""
    return auto_scaler


def get_health_checker() -> Optional[HealthChecker]:
    """Get the global health checker instance."""
    return health_checker


def get_cluster_controller() -> Optional[DistributedController]:
    """Get the global cluster controller instance."""
    return cluster_controller


def get_nginx_tracking():
    """Get nginx request tracking state."""
    global _prev_nginx_requests, _prev_nginx_time
    return _prev_nginx_requests, _prev_nginx_time


def set_nginx_tracking(requests: Optional[int], timestamp: Optional[float]):
    """Set nginx request tracking state."""
    global _prev_nginx_requests, _prev_nginx_time
    _prev_nginx_requests = requests
    _prev_nginx_time = timestamp


def on_become_leader():
    """Called when this node becomes the cluster leader"""
    logger.info("👑 This node has become the cluster leader - taking control of operations")

    if app_manager and auto_scaler:
        try:
            adopted_summary = app_manager.reconcile_all()
            logger.info(f"✅ Leader reconciled existing containers: {adopted_summary}")
        except Exception as e:
            logger.error(f"❌ Leader failed to reconcile existing containers: {e}")
        
        try:
            apps = state_store.list_apps()
            logger.info(f"🔄 Restoring scaling policies for {len(apps)} apps from database")
            
            for app in apps:
                app_name = app["name"]
                try:
                    # Get full app record to access the spec with scaling config
                    app_record = state_store.get_app(app_name)
                    if app_record and app_record.spec:
                        scaling_config = app_record.spec.get("scaling", {})
                        
                        if scaling_config:  # Only restore if scaling config exists
                            policy = ScalingPolicy(
                                min_replicas=scaling_config["minReplicas"],
                                max_replicas=scaling_config["maxReplicas"],
                                target_rps_per_replica=scaling_config["targetRPSPerReplica"],
                                max_p95_latency_ms=scaling_config["maxP95LatencyMs"],
                                scale_out_threshold_pct=scaling_config["scaleOutThresholdPct"],
                                scale_in_threshold_pct=scaling_config["scaleInThresholdPct"],
                                window_seconds=scaling_config.get("windowSeconds", 60),  # Optional field
                                cooldown_seconds=scaling_config.get("cooldownSeconds", 300)  # Optional field
                            )
                            
                            auto_scaler.set_policy(app_name, policy)
                            logger.info(f"✅ Restored scaling policy for {app_name}: targetRPS={scaling_config['targetRPSPerReplica']}, thresholds={scaling_config['scaleOutThresholdPct']}%/{scaling_config['scaleInThresholdPct']}%")
                        else:
                            logger.debug(f"No scaling config found in spec for {app_name}")
                    else:
                        logger.warning(f"Could not get app record for {app_name}")
                        
                except Exception as e:
                    logger.error(f"❌ Failed to restore scaling policy for {app_name}: {e}")
                    
            logger.info("✅ Leader completed scaling policy restoration from database")
        except Exception as e:
            logger.error(f"❌ Leader failed to restore scaling policies: {e}")
        
        # Start container monitoring for automatic restarts and minReplicas enforcement
        app_manager.start_container_monitoring()
        
        # Clean only containers whose app spec no longer exists
        try:
            app_manager.cleanup_orphaned_containers()
            logger.info("✅ Leader completed orphaned container cleanup")
        except Exception as e:
            logger.error(f"❌ Leader failed orphaned container cleanup: {e}")


def on_lose_leadership():
    """Called when this node loses leadership"""
    logger.warning("💔 This node has lost cluster leadership - stepping down from operations")
    # Follower nodes should stop active operations and just serve read requests
    if app_manager:
        app_manager.stop_container_monitoring()


def on_cluster_change(nodes):
    """Called when cluster membership changes"""
    node_count = len(nodes)
    node_ids = [node.node_id for node in nodes.values()]
    logger.info(f"🔄 Cluster membership changed: {node_count} nodes - {node_ids}")


def background_monitoring():
    """Background thread for monitoring and autoscaling."""
    logger.info("Started background monitoring thread")
    
    while monitoring_active:
        try:
            if not app_manager or not auto_scaler:
                time.sleep(5)
                continue
                
            # Only run monitoring on the leader node
            if cluster_controller and not cluster_controller.is_leader:
                time.sleep(5)
                continue
            
            # Get list of running apps only - don't scale stopped apps
            all_apps = state_store.list_apps()
            apps = [app for app in all_apps if app.get("status") == "running"]

            # Fetch nginx status once per loop for reuse
            nginx_status = nginx_manager.get_nginx_status() if nginx_manager else None
            
            for app_info in apps:
                app_name = app_info["name"]
                
                try:
                    # Use the scaler to observe metrics and make scaling decisions
                    auto_scaler.observe(app_name, nginx_status)
                    
                    decision = auto_scaler.decide(app_name)
                    
                    if decision and decision.action != "none":
                        logger.info(f"Scaling decision for {app_name}: {decision.action} to {decision.target_replicas} replicas. Reason: {decision.reason}")
                        
                        try:
                            if decision.action == "scale_out":
                                app_manager.scale_app(app_name, decision.target_replicas)
                            elif decision.action == "scale_in":
                                app_manager.scale_app(app_name, decision.target_replicas)
                        except Exception as e:
                            logger.error(f"Failed to execute scaling action for {app_name}: {e}")
                
                except Exception as e:
                    logger.error(f"Error during monitoring loop for {app_name}: {e}")
            
            # Sleep between monitoring cycles
            time.sleep(5)
            
        except Exception as e:
            logger.error(f"Error in background monitoring thread: {e}")
            time.sleep(5)
    
    logger.info("Background monitoring thread stopped")


async def startup_event():
    """Initialize all components when the API starts."""
    global app_manager, state_store, nginx_manager, auto_scaler, health_checker, cluster_controller
    global monitoring_task, monitoring_active
    
    try:
        # Initialize PostgreSQL High Availability database cluster
        logger.info("🚀 Initializing PostgreSQL HA database cluster...")
        state_store = get_database_manager()
        
        # Initialize distributed controller cluster with leader election
        logger.info("🏗️  Initializing distributed controller cluster...")
        cluster_controller = DistributedController(
            node_id=os.getenv("CLUSTER_NODE_ID"),
            hostname=os.getenv("CLUSTER_HOSTNAME", "localhost"),
            port=int(os.getenv("ORCHESTRY_PORT", "8000")),
            db_manager=state_store
        )
        
        # Set up cluster event handlers
        cluster_controller.on_become_leader = on_become_leader
        cluster_controller.on_lose_leadership = on_lose_leadership
        cluster_controller.on_cluster_change = on_cluster_change
        
        # Start the cluster
        cluster_controller.start()
        
        # Initialize other components
        nginx_manager = DockerNginxManager()
        auto_scaler = AutoScaler()
        health_checker = HealthChecker()
        app_manager = AppManager(state_store, nginx_manager)
        
        # Start health checker
        await health_checker.start()
        
        # Reconcile existing containers BEFORE cleanup
        try:
            adopted_summary = app_manager.reconcile_all()
            logger.info(f"Reconciliation summary on startup: {adopted_summary}")
            
            # Re-register scaling policies for all existing apps from database
            logger.info("Restoring scaling policies from database on startup")
            apps = state_store.list_apps()
            logger.info(f"Found {len(apps)} apps to restore scaling policies for")
            
            for app in apps:
                app_name = app["name"]
                logger.info(f"Attempting to restore scaling policy for {app_name}")
                try:
                    # Get full app record to access the spec with scaling config
                    app_record = state_store.get_app(app_name)
                    if app_record and app_record.spec:
                        scaling_config = app_record.spec.get("scaling", {})
                        
                        if scaling_config:  # Only restore if scaling config exists
                            logger.info(f"Scaling config for {app_name}: {scaling_config}")
                            
                            policy = ScalingPolicy(
                                min_replicas=scaling_config["minReplicas"],
                                max_replicas=scaling_config["maxReplicas"],
                                target_rps_per_replica=scaling_config["targetRPSPerReplica"],
                                max_p95_latency_ms=scaling_config["maxP95LatencyMs"],
                                scale_out_threshold_pct=scaling_config["scaleOutThresholdPct"],
                                scale_in_threshold_pct=scaling_config["scaleInThresholdPct"],
                                window_seconds=scaling_config.get("windowSeconds", 60),  # Optional field
                                cooldown_seconds=scaling_config.get("cooldownSeconds", 300)  # Optional field
                            )
                            
                            auto_scaler.set_policy(app_name, policy)
                            logger.info(f"Successfully restored scaling policy for {app_name}: targetRPS={scaling_config['targetRPSPerReplica']}, thresholds={scaling_config['scaleOutThresholdPct']}%/{scaling_config['scaleInThresholdPct']}%")
                        else:
                            logger.debug(f"No scaling config found in spec for {app_name}")
                    else:
                        logger.warning(f"Could not get app record for {app_name}")
                        
                except Exception as e:
                    logger.error(f"Failed to restore scaling policy for {app_name}: {e}")
                    import traceback
                    logger.error(f"Full traceback: {traceback.format_exc()}")
                    
        except Exception as e:
            logger.error(f"Failed initial reconciliation: {e}")
            import traceback
            logger.error(f"Full traceback: {traceback.format_exc()}")

        # Start background monitoring (runs on all nodes but only leader does work)
        monitoring_active = True
        monitoring_task = threading.Thread(target=background_monitoring, daemon=True)
        monitoring_task.start()
        
        # Wait a bit for cluster to elect leader, then do cleanup
        # Container monitoring and cleanup will be started by the leader
        logger.info("Waiting for leader election before starting container operations...")
        
        logger.info("Orchestry Controller API started successfully")
        
    except Exception as e:
        logger.error(f"Failed to start controller: {e}")
        raise


async def shutdown_event():
    """Clean up resources when shutting down."""
    global monitoring_active, health_checker, app_manager, cluster_controller
    
    monitoring_active = False
    
    if cluster_controller:
        cluster_controller.stop()
    
    if app_manager:
        app_manager.stop_container_monitoring()
    
    if health_checker:
        await health_checker.stop()
    
    if state_store:
        state_store.close()
    
    logger.info("Orchestry Controller API shut down")

File: controller_go/lifecycle.go
package controller

import (
	"context"
	"log"
	"os"
	"strconv"
	"sync"
	"time"

	"statego"
)

// Lifecycle manages the startup and shutdown of all controller components
type Lifecycle struct {
	// Components
	appManager        *AppManager
	stateStore        *PostgresStateStore
	nginxManager      *NginxManagerImpl
	autoScaler        *AutoScaler
	healthChecker     *HealthChecker
	clusterController *DistributedController
	apiServer         *APIServer

	// Monitoring state
	monitoringActive bool
	monitoringCancel context.CancelFunc
	monitoringCtx    context.Context
	monitoringWg     sync.WaitGroup

	// Nginx tracking for RPS calculation
	prevNginxRequests  *int
	prevNginxTime      *float64
	nginxTrackingMutex sync.Mutex
}

// NewLifecycle creates a new lifecycle manager
func NewLifecycle() *Lifecycle {
	return &Lifecycle{}
}

// Startup initializes all components
func (l *Lifecycle) Startup() error {
	log.Println("🚀 Starting Orchestry Controller...")

	// 1. Initialize PostgreSQL HA database cluster
	log.Println("🚀 Initializing PostgreSQL HA database cluster...")

	// Get database configuration from environment
	primaryHost := os.Getenv("POSTGRES_PRIMARY_HOST")
	if primaryHost == "" {
		primaryHost = "postgres-primary"
	}

	primaryPort := 5432
	if portStr := os.Getenv("POSTGRES_PRIMARY_PORT"); portStr != "" {
		if p, err := strconv.Atoi(portStr); err == nil {
			primaryPort = p
		}
	}

	replicaHost := os.Getenv("POSTGRES_REPLICA_HOST")
	if replicaHost == "" {
		replicaHost = "postgres-replica"
	}

	replicaPort := 5432
	if portStr := os.Getenv("POSTGRES_REPLICA_PORT"); portStr != "" {
		if p, err := strconv.Atoi(portStr); err == nil {
			replicaPort = p
		}
	}

	database := os.Getenv("POSTGRES_DB")
	if database == "" {
		database = "orchestry"
	}

	username := os.Getenv("POSTGRES_USER")
	if username == "" {
		username = "orchestry"
	}

	password := os.Getenv("POSTGRES_PASSWORD")
	if password == "" {
		password = "orchestry_password"
	}

	minConn := 5
	if connStr := os.Getenv("POSTGRES_MIN_CONNECTIONS"); connStr != "" {
		if c, err := strconv.Atoi(connStr); err == nil {
			minConn = c
		}
	}

	maxConn := 20
	if connStr := os.Getenv("POSTGRES_MAX_CONNECTIONS"); connStr != "" {
		if c, err := strconv.Atoi(connStr); err == nil {
			maxConn = c
		}
	}

	dbManager, err := statego.NewDatabaseManager(
		primaryHost, primaryPort,
		replicaHost, replicaPort,
		database, username, password,
		minConn, maxConn,
	)
	if err != nil {
		return err
	}
	l.stateStore = NewPostgresStateStore(dbManager)

	// 2. Initialize distributed controller cluster with leader election
	log.Println("🏗️  Initializing distributed controller cluster...")
	nodeID := os.Getenv("CLUSTER_NODE_ID")
	hostname := os.Getenv("CLUSTER_HOSTNAME")
	if hostname == "" {
		hostname = "localhost"
	}

	portStr := os.Getenv("ORCHESTRY_PORT")
	if portStr == "" {
		portStr = "8000"
	}
	port, err := strconv.Atoi(portStr)
	if err != nil {
		return err
	}

	// Cluster configuration parameters
	leaseTTL := 10         // seconds
	heartbeatInterval := 3 // seconds
	electionTimeout := 5   // seconds

	l.clusterController, err = NewDistributedController(
		nodeID, hostname, port, dbManager,
		leaseTTL, heartbeatInterval, electionTimeout,
	)
	if err != nil {
		return err
	}

	// Set up cluster event handlers
	l.clusterController.onBecomeLeader = l.onBecomeLeader
	l.clusterController.onLoseLeadership = l.onLoseLeadership
	l.clusterController.onClusterChange = l.onClusterChange

	// Start the cluster
	if err := l.clusterController.Start(); err != nil {
		return err
	}

	// 3. Initialize other components
	nginxContainer := os.Getenv("ORCHESTRY_NGINX_CONTAINER")
	nginxConfDir := os.Getenv("ORCHESTRY_NGINX_CONF_DIR")
	l.nginxManager, err = NewNginxManager(nginxContainer, nginxConfDir, "")
	if err != nil {
		return err
	}

	l.autoScaler = NewAutoScaler()
	l.healthChecker = NewHealthChecker()

	l.appManager, err = NewAppManager(l.stateStore, l.nginxManager)
	if err != nil {
		return err
	}

	// Set health checker in app manager
	l.appManager.healthChecker = l.healthChecker

	// 4. Start health checker
	l.healthChecker.Start()

	// 5. Reconcile existing containers BEFORE cleanup
	log.Println("🔄 Reconciling existing containers...")
	adoptedSummary := l.appManager.ReconcileApp("")
	log.Printf("Reconciliation summary on startup: adopted %d containers", adoptedSummary)

	// 6. Restore scaling policies from database
	log.Println("📋 Restoring scaling policies from database...")
	if err := l.restoreScalingPolicies(); err != nil {
		log.Printf("Warning: Failed to restore some scaling policies: %v", err)
	}

	// 7. Create API server
	l.apiServer, err = NewAPIServer(
		l.appManager,
		l.stateStore,
		l.nginxManager,
		l.autoScaler,
		l.healthChecker,
		l.clusterController,
	)
	if err != nil {
		return err
	}

	// 8. Start background monitoring
	l.monitoringCtx, l.monitoringCancel = context.WithCancel(context.Background())
	l.monitoringActive = true
	l.monitoringWg.Add(1)
	go l.backgroundMonitoring()

	log.Println("✅ Orchestry Controller started successfully")
	return nil
}

// Shutdown cleans up all resources
func (l *Lifecycle) Shutdown() {
	log.Println("🛑 Shutting down Orchestry Controller...")

	// Stop monitoring
	l.monitoringActive = false
	if l.monitoringCancel != nil {
		l.monitoringCancel()
	}
	l.monitoringWg.Wait()

	// Stop cluster
	if l.clusterController != nil {
		l.clusterController.Stop()
	}

	// Stop container monitoring
	if l.appManager != nil {
		l.appManager.StopContainerMonitoring()
	}

	// Stop health checker
	if l.healthChecker != nil {
		l.healthChecker.Stop()
	}

	// Close database connections
	if l.stateStore != nil && l.stateStore.dbManager != nil {
		// Implement close if needed
	}

	log.Println("✅ Orchestry Controller shut down")
}

// Run starts the API server (blocking call)
func (l *Lifecycle) Run() error {
	host := os.Getenv("ORCHESTRY_HOST")
	if host == "" {
		host = "0.0.0.0"
	}

	portStr := os.Getenv("ORCHESTRY_PORT")
	if portStr == "" {
		portStr = "8000"
	}

	addr := host + ":" + portStr
	return l.apiServer.Run(addr)
}

// onBecomeLeader is called when this node becomes the cluster leader
func (l *Lifecycle) onBecomeLeader() {
	log.Println("👑 This node has become the cluster leader - taking control of operations")

	// Reconcile existing containers
	if l.appManager != nil && l.autoScaler != nil {
		adoptedSummary := l.appManager.ReconcileApp("")
		log.Printf("✅ Leader reconciled existing containers: %d adopted", adoptedSummary)

		// Restore scaling policies
		if err := l.restoreScalingPolicies(); err != nil {
			log.Printf("❌ Leader failed to restore scaling policies: %v", err)
		}

		// Start container monitoring
		l.appManager.StartContainerMonitoring()

		// Cleanup orphaned containers
		// Note: CleanupOrphanedContainers method needs to be implemented in manager.go
		log.Println("✅ Leader completed setup")
	}
}

// onLoseLeadership is called when this node loses leadership
func (l *Lifecycle) onLoseLeadership() {
	log.Println("💔 This node has lost cluster leadership - stepping down from operations")

	if l.appManager != nil {
		l.appManager.StopContainerMonitoring()
	}
}

// onClusterChange is called when cluster membership changes
func (l *Lifecycle) onClusterChange(nodes map[string]*ClusterNode) {
	nodeCount := len(nodes)
	nodeIDs := make([]string, 0, nodeCount)
	for _, node := range nodes {
		nodeIDs = append(nodeIDs, node.NodeID)
	}
	log.Printf("🔄 Cluster membership changed: %d nodes - %v", nodeCount, nodeIDs)
}

// restoreScalingPolicies restores scaling policies from the database
func (l *Lifecycle) restoreScalingPolicies() error {
	apps, err := l.stateStore.ListApps()
	if err != nil {
		return err
	}

	log.Printf("🔄 Restoring scaling policies for %d apps from database", len(apps))

	for _, app := range apps {
		appName := app.Name

		// Get full app record to access the spec with scaling config
		appRecord, err := l.stateStore.GetApp(appName)
		if err != nil || appRecord == nil {
			log.Printf("Warning: Could not get app record for %s", appName)
			continue
		}

		// Extract scaling config from spec
		scalingConfig, ok := appRecord.Spec["scaling"].(map[string]interface{})
		if !ok || scalingConfig == nil {
			log.Printf("No scaling config found for %s", appName)
			continue
		}

		// Create scaling policy
		policy := ScalingPolicy{
			MinReplicas:          getIntFromMapDefault(scalingConfig, "minReplicas", 1),
			MaxReplicas:          getIntFromMapDefault(scalingConfig, "maxReplicas", 5),
			TargetRPSPerReplica:  getIntFromMapDefault(scalingConfig, "targetRPSPerReplica", 50),
			MaxP95LatencyMs:      getIntFromMapDefault(scalingConfig, "maxP95LatencyMs", 250),
			ScaleOutThresholdPct: getIntFromMapDefault(scalingConfig, "scaleOutThresholdPct", 80),
			ScaleInThresholdPct:  getIntFromMapDefault(scalingConfig, "scaleInThresholdPct", 30),
			WindowSeconds:        getIntFromMapDefault(scalingConfig, "windowSeconds", 60),
			CooldownSeconds:      getIntFromMapDefault(scalingConfig, "cooldownSeconds", 300),
		}

		if err := l.autoScaler.SetPolicy(appName, policy); err != nil {
			log.Printf("❌ Failed to restore scaling policy for %s: %v", appName, err)
		} else {
			log.Printf("✅ Restored scaling policy for %s: targetRPS=%d, thresholds=%d%%/%d%%",
				appName, policy.TargetRPSPerReplica, policy.ScaleOutThresholdPct, policy.ScaleInThresholdPct)
		}
	}

	log.Println("✅ Completed scaling policy restoration from database")
	return nil
}

// backgroundMonitoring runs the monitoring and autoscaling loop
func (l *Lifecycle) backgroundMonitoring() {
	defer l.monitoringWg.Done()
	log.Println("Started background monitoring thread")

	ticker := time.NewTicker(10 * time.Second)
	defer ticker.Stop()

	for {
		select {
		case <-l.monitoringCtx.Done():
			log.Println("Background monitoring thread stopped")
			return
		case <-ticker.C:
			if err := l.monitoringCycle(); err != nil {
				log.Printf("Error in background monitoring: %v", err)
			}
		}
	}
}

// monitoringCycle performs one monitoring and autoscaling cycle
func (l *Lifecycle) monitoringCycle() error {
	// Only run monitoring on the leader node
	if l.clusterController != nil && !l.clusterController.isLeader {
		return nil
	}

	// Get list of running apps only
	allApps, err := l.stateStore.ListApps()
	if err != nil {
		return err
	}

	// Filter to only running apps
	runningApps := []*AppRecord{}
	for _, app := range allApps {
		if app.Status == "running" {
			runningApps = append(runningApps, app)
		}
	}

	// Fetch nginx status for global metrics
	nginxStatus, err := l.nginxManager.GetNginxStatus()
	if err != nil {
		log.Printf("Warning: Unable to fetch nginx status: %v", err)
		nginxStatus = map[string]interface{}{}
	}

	// Compute global RPS from nginx
	rpsGlobal := l.computeGlobalRPS(nginxStatus)
	activeConnsGlobal := getIntFromMapDefault(nginxStatus, "active_connections", 0)

	// Calculate total replicas across all apps for fair-share metrics
	totalReplicasGlobal := 0
	for _, app := range runningApps {
		if instances, ok := l.appManager.instances[app.Name]; ok {
			totalReplicasGlobal += len(instances)
		}
	}

	// Process each running app
	for _, appInfo := range runningApps {
		appName := appInfo.Name

		// Get current instances
		instances, ok := l.appManager.instances[appName]
		if !ok || len(instances) == 0 {
			continue
		}

		// Update container stats
		l.appManager.UpdateContainerStats(appName)

		// Collect metrics
		healthyCount := 0
		totalCPU := 0.0
		totalMemory := 0.0

		for _, inst := range instances {
			if inst.State == "ready" {
				healthyCount++
			}
			totalCPU += inst.CPUPercent
			totalMemory += inst.MemoryPercent
		}

		avgCPU := totalCPU / float64(len(instances))
		avgMemory := totalMemory / float64(len(instances))

		// Fair-share distribution of global RPS & connections
		share := 0.0
		if totalReplicasGlobal > 0 {
			share = float64(len(instances)) / float64(totalReplicasGlobal)
		}
		appRPS := rpsGlobal * share
		appActiveConns := int(float64(activeConnsGlobal) * share)

		metrics := ScalingMetrics{
			RPS:               appRPS,
			P95LatencyMs:      0, // Not implemented yet
			ActiveConnections: appActiveConns,
			CPUPercent:        avgCPU,
			MemoryPercent:     avgMemory,
			HealthyReplicas:   healthyCount,
			TotalReplicas:     len(instances),
		}

		// Add metrics to scaler
		l.autoScaler.AddMetrics(appName, metrics)

		// Get app mode
		appRecord, _ := l.stateStore.GetApp(appName)
		appMode := "auto"
		if appRecord != nil {
			appMode = appRecord.Mode
		}

		// Evaluate scaling decision
		decision := l.autoScaler.Evaluate(appName, len(instances), appMode)

		// Log evaluation (useful for debugging)
		policy := l.autoScaler.GetPolicy(appName)
		if policy != nil {
			log.Printf("Scaling evaluation for %s: RPS=%.2f, Conns=%d, CPU=%.1f%%, Mem=%.1f%%, "+
				"Replicas=%d, Decision=%v, Reason=%s",
				appName, metrics.RPS, metrics.ActiveConnections, avgCPU, avgMemory,
				len(instances), decision.ShouldScale, decision.Reason)
		}

		// Execute scaling if needed
		if decision.ShouldScale {
			log.Printf("Scaling %s: %s", appName, decision.Reason)

			result := l.appManager.Scale(appName, decision.TargetReplicas)

			if result["status"] == "scaled" {
				// Record scaling action
				l.autoScaler.RecordScalingAction(appName, decision.TargetReplicas)

				// Log to state store
				l.stateStore.LogScalingAction(
					appName,
					decision.CurrentReplicas,
					decision.TargetReplicas,
					decision.Reason,
					decision.TriggeredBy,
					&decision.Metrics,
				)

				// Log event
				l.stateStore.LogEvent(appName, "scaled", map[string]interface{}{
					"old_replicas": decision.CurrentReplicas,
					"new_replicas": decision.TargetReplicas,
					"reason":       decision.Reason,
				})
			}
		}
	}

	return nil
}

// computeGlobalRPS computes RPS from nginx status
func (l *Lifecycle) computeGlobalRPS(nginxStatus map[string]interface{}) float64 {
	l.nginxTrackingMutex.Lock()
	defer l.nginxTrackingMutex.Unlock()

	currentRequests, ok := nginxStatus["requests"].(int)
	if !ok {
		return 0.0
	}

	now := float64(time.Now().Unix())

	if l.prevNginxRequests != nil && l.prevNginxTime != nil {
		deltaReq := currentRequests - *l.prevNginxRequests
		deltaTime := now - *l.prevNginxTime

		if deltaReq >= 0 && deltaTime > 0 {
			rps := float64(deltaReq) / deltaTime
			l.prevNginxRequests = &currentRequests
			l.prevNginxTime = &now
			return rps
		}
	}

	// Initialize tracking
	l.prevNginxRequests = &currentRequests
	l.prevNginxTime = &now
	return 0.0
}

// Helper function
func getIntFromMapDefault(m map[string]interface{}, key string, defaultVal int) int {
	if val, ok := m[key]; ok {
		switch v := val.(type) {
		case int:
			return v
		case float64:
			return int(v)
		case string:
			if i, err := strconv.Atoi(v); err == nil {
				return i
			}
		}
	}
	return defaultVal
}

// UpdateContainerStats updates CPU and memory stats for containers
func (am *AppManager) UpdateContainerStats(appName string) {
	// This would fetch stats from Docker
	// Simplified for now - implement full stats gathering as needed
}
